### ISSUE TITLE
Slow running unit tests

### ISSUE THREAD

**Issue by neolit123** [kubernetes/kubeadm#952]
@fabriziopandini has reported some slow running unit tests:
```
//cmd/kubeadm/app/cmd/phases:go_default_test                             PASSED in 44.8s
//cmd/kubeadm/app/discovery/token:go_default_test                        PASSED in 19.3s
//cmd/kubeadm/app/phases/certs:go_default_test                           PASSED in 34.1s
//cmd/kubeadm/app/phases/certs/pkiutil:go_default_test                   PASSED in 7.1s
//cmd/kubeadm/app/phases/controlplane:go_default_test                    PASSED in 5.4s
//cmd/kubeadm/app/phases/kubeconfig:go_default_test                      PASSED in 12.9s
//cmd/kubeadm/app/phases/markmaster:go_default_test                      PASSED in 3.7s
//cmd/kubeadm/app/phases/upgrade:go_default_test                         PASSED in 18.2s
```

in my case tests run significantly faster - e.g. `cmd/kubeadm/app/cmd/phase` takes around ~20 seconds.
but we are logging this issue in case this can be improved.

for instance i did some investigation in here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/certs_test.go#L105
(which is one of the longest running tests).

i don't see anything wrong with the code, which to me means that the code simply does a lot of operations (too many files + sub-commands?). we could either remove the amount of tests or leave it as is for now.

**Comment by timothysc** [comment]
/assign @liztio   

It really seems the cert + token tests are indeed slow.

**Comment by timothysc** [comment]
@liztio - This crosses over https://github.com/kubernetes/kubernetes/issues/64625 and is visible on own CI.  

Bumping priority.

**Comment by ncdc** [comment]
@timothysc I don't see `//cmd/kubeadm/app/cmd` in the list at the top (which is where `TestRunListTokens` lives), so it might not be 100% related, but thanks for the cc!

**Comment by neolit123** [comment]
`TestRunListTokens` is for me to blame and yes i saw the flakyness at some point....
running a server to test token CLI is really a bad ideas in the first place and we should probably just nuke it or move to integration tests.

**Comment by neolit123** [comment]
@bart0sh please, don't reference issues in commit bodies as this can spam a bit.

**Comment by ncdc** [comment]
@neolit123 I've been looking into it and when I see the 5 minute timeout, it's in this block of code: https://github.com/kubernetes/kubernetes/blob/17c77c7898218073f14c8d573582e8d2313dc740/cmd/kubeadm/app/cmd/token_test.go#L448-L476.

Here is the output from the test run:
```
=== RUN   TestRunListTokens
[dryrun] Would perform action LIST on resource "secrets" in API group "core/v1"
dummy API server listening on localhost:9008
Running HTTP test case (0) "token-id not defined"
[dryrun] Would perform action LIST on resource "secrets" in API group "core/v1"
panic: test timed out after 5m0s

goroutine 15 [running]:
testing.(*M).startAlarm.func1()
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:1240 +0xfc
created by time.goFunc
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/time/sleep.go:172 +0x44

goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0xc420711590, 0x166ce02, 0x11, 0x1705748, 0x47fb01)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:825 +0x301
testing.runTests.func1(0xc420518870)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:1063 +0x64
testing.tRunner(0xc420518870, 0xc4204c5df8)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:777 +0xd0
testing.runTests(0xc420399640, 0x2179b40, 0x14, 0x14, 0x411fb9)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:1061 +0x2c4
testing.(*M).Run(0xc4203be180, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:978 +0x171
main.main()
	_testmain.go:82 +0x151

goroutine 19 [chan receive]:
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x2188940)
	/home/ubuntu/workspace/hks-kubernetes-unit/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x8b
created by k8s.io/kubernetes/vendor/github.com/golang/glog.init.0
	/home/ubuntu/workspace/hks-kubernetes-unit/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:410 +0x203

goroutine 35 [syscall, 5 minutes]:
os/signal.signal_recv(0xc4202c1fc8)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/os/signal/signal_unix.go:28 +0x41

goroutine 52 [IO wait, 4 minutes]:
internal/poll.runtime_pollWait(0x7f81b7e16f00, 0x72, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42081a018, 0x72, 0xc4203c8400, 0x0, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc42081a018, 0xffffffffffffff00, 0x0, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc42081a000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/internal/poll/fd_unix.go:372 +0x1a8
net.(*netFD).accept(0xc42081a000, 0xc420812180, 0x0, 0xc000000000)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc420818008, 0x1790830, 0x6, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).AcceptTCP(0xc420818008, 0x1573540, 0xc420812180, 0x14704c0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/tcpsock.go:246 +0x49
net/http.tcpKeepAliveListener.Accept(0xc420818008, 0xc4200d6050, 0x14704c0, 0x216c100, 0x161aa20)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/http/server.go:3219 +0x2f
net/http.(*Server).Serve(0xc420774750, 0x17be900, 0xc420818008, 0x0, 0x0)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/http/server.go:2773 +0x1a5
net/http.(*Server).ListenAndServe(0xc420774750, 0xc42077d272, 0xe)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/net/http/server.go:2714 +0xa9
k8s.io/kubernetes/cmd/kubeadm/app/cmd.TestRunListTokens.func1(0xc420774750, 0xc420711590, 0x165de4e, 0x4)
	/home/ubuntu/workspace/hks-kubernetes-unit/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/token_test.go:395 +0x2f
created by k8s.io/kubernetes/cmd/kubeadm/app/cmd.TestRunListTokens
	/home/ubuntu/workspace/hks-kubernetes-unit/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/token_test.go:394 +0x6a6

goroutine 32 [sleep]:
time.Sleep(0x2faf080)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/runtime/time.go:102 +0x166
k8s.io/kubernetes/cmd/kubeadm/app/cmd.TestRunListTokens(0xc420711590)
	/home/ubuntu/workspace/hks-kubernetes-unit/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/token_test.go:469 +0xa5b
testing.tRunner(0xc420711590, 0x1705748)
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
	/home/ubuntu/.gimme/versions/go1.10.4.linux.amd64/src/testing/testing.go:824 +0x2e0
FAIL	k8s.io/kubernetes/cmd/kubeadm/app/cmd	300.052s
```

Looking at the running goroutines, it looks like it’s only made it to test case 0, and is somehow stuck in that for loop sleeping, even though it’s supposed to be capped at 2000 ms?

**Comment by neolit123** [comment]
it's odd, i need to debug it and found the reason.
just ran the test 3 times in a row without a problem.

**Comment by ncdc** [comment]
Yeah, I can run it by itself hundreds of times without issue. I think it must be contention when you run `make test`, which compiles/links/tests in parallel up to the # of cpus you have.

**Comment by neolit123** [comment]
concurrency might be an issue here, but also this:
```
	http.HandleFunc("/", httpHandler)
```
that's a pretty explicit handler.

**Comment by ncdc** [comment]
Right - I am talking about concurrency with other go packages in kubernetes, as they're tested in parallel.

**Comment by neolit123** [comment]
it's odd. i cannot reproduce this with `go test` or `make test`, but i have definitely seen this on this machine.

my goto idea at this point is just to remove the test based on these points:
- it's not a CLI unit test.
- it's testing API objects using a fake API server which is a bad hack...
- existing e2e tests already provide coverage for all the `expectedError: false` cases.

filing a PR later today.

**Comment by ncdc** [comment]
FYI I've seen it somewhat frequently in AWS on an `m5.xlarge` instance. 👍 to removing it though 😄

**Comment by liztio** [comment]
I can't repro the flakes. Running with `--runs_per_test 10` produces no failures.

My slowest tests were:

```
//cmd/kubeadm/test/cmd:go_default_test                                   PASSED in 46.8s
  Stats over 10 runs: max = 46.8s, min = 39.6s, avg = 44.0s, dev = 2.1s
//cmd/kubeadm/app/phases/certs:go_default_test                           PASSED in 39.2s
  Stats over 10 runs: max = 39.2s, min = 32.5s, avg = 36.4s, dev = 2.2s
//cmd/kubeadm/app/phases/upgrade:go_default_test                         PASSED in 27.4s
  Stats over 10 runs: max = 27.4s, min = 23.6s, avg = 24.8s, dev = 1.2s
//cmd/kubeadm/app/cmd/phases/certs:go_default_test                       PASSED in 18.0s
  Stats over 10 runs: max = 18.0s, min = 12.8s, avg = 15.9s, dev = 1.6s
```

**Comment by neolit123** [comment]
PR to remove TestRunListTokens sent :+1: 
https://github.com/kubernetes/kubernetes/pull/70452

this tracking issue is about overall slow running unit tests in kubeadm.

**Comment by timothysc** [comment]
Critical one is merged punting the rest to 1.14

**Comment by neolit123** [comment]
i'm running the tests on a regular basis before a PR and i don't think the tests are *that* slow.
if someone thinks this is still a problem, please open a new ticket with precise details outlining problematic tests.
